weave/trace/op.py

```diff
@@ -256,3 +62,264 @@ def _apply_fn_defaults_to_inputs(
+    except TypeError as e:
+        raise OpCallError(f"Error calling {func.name}: {e}")
+    inputs_with_defaults = _apply_fn_defaults_to_inputs(func, inputs)
+
+    # This should probably be configurable, but for now we redact the api_key
+    if "api_key" in inputs_with_defaults:
+        inputs_with_defaults["api_key"] = "REDACTED"
+
+    # If/When we do memoization, this would be a good spot
+
+    parent_call = call_context.get_current_call()
+    client._save_nested_objects(inputs_with_defaults)
+    attributes = call_attributes.get()
+
+    return client.create_call(
+        func,
+        inputs_with_defaults,
+        parent_call,
+        attributes=attributes,
+    )
+
+
+def _execute_call(
+    __op: Op,
+    call: Any,
+    *args: Any,
+    __return_type: Literal["call", "value"] = "call",
+    **kwargs: Any,
+) -> Any:
+    func = __op.resolve_fn
+    client = client_context.weave_client.require_weave_client()
+    has_finished = False
+
+    def finish(output: Any = None, exception: Optional[BaseException] = None) -> None:
+        nonlocal has_finished
+        if has_finished:
+            raise ValueError("Should not call finish more than once")
+        client.finish_call(call, output, exception)
+        if not call_context.get_current_call():
+            print_call_link(call)
+
+    def on_output(output: Any) -> Any:
+        if handler := getattr(__op, "_on_output_handler", None):
+            return handler(output, finish, call.inputs)
+        finish(output)
+        return output
+
+    try:
+        res = func(*args, **kwargs)
+    except BaseException as e:
+        finish(exception=e)
+        raise
+    else:
+        res = box.box(res)  # TODO: can we get rid of this?
+    # We cannot let BoxedNone or BoxedBool escape into the user's code
+    # since they cannot pass instance checks for None or bool.
+    if isinstance(res, box.BoxedNone):
+        res = None
+    if isinstance(res, box.BoxedBool):
+        res = res.val
+
+    if inspect.iscoroutine(res):
+        awaitable = res
+
+        async def _call_async() -> Coroutine[Any, Any, Any]:
+            try:
+                call_context.push_call(call)
+                output = await awaitable
+                res2 = on_output(output)
+                return call if __return_type == "call" else res2
+            except BaseException as e:
+                finish(exception=e)
+                raise
+            finally:
+                call_context.pop_call(call.id)
+
+        return _call_async()
+    else:
+        res2 = on_output(res)
+        return call if __return_type == "call" else res2
+
+
+def call(op: Op, *args: Any, **kwargs: Any) -> Any:
+    c = _create_call(op, *args, **kwargs)
+    return _execute_call(op, c, *args, **kwargs)
+
+
+def calls(op: Op) -> "CallsIter":
+    client = client_context.weave_client.require_weave_client()
+    return client._op_calls(op)
+
+
+# Legacy decos
```

Comment:
What is legacy decos? We don't need this. legacy.decorator_op has the legacy decorator, but we shouldn't need to redefine anything here.

Reply:
would it be possible to put these in a different file and selectively import??

Reply:
we can try, but let's do that in a different PR.  This has enough dangerous changes as is!